ark/utils/_bootstrapping.pyx

```diff
@@ -0,0 +1,276 @@
+from cython cimport cdivision, boundscheck, wraparound
+from cpython.mem cimport PyMem_Malloc, PyMem_Free
+
+import numpy as np
+cimport numpy as np
+
+from libc.stdlib cimport rand
+from libc.string cimport memset
+
+ctypedef np.uint64_t MAXINDEX_t
+ctypedef np.uint16_t DTYPE_t
+ctypedef np.uint8_t UINT8_t
+
+@cdivision(True) # Ignore modulo/divide by zero warning
+cdef inline void _c_permutation(Py_ssize_t* arr, const Py_ssize_t size) nogil:
+    """ Randomly permutes the provided arr
+
+    Very common implementation
+
+    Args:
+        arr (Py_ssize_t*):
+            array to be permuted/shuffled
+        size (Py_ssize_t):
+            size of array to be shuffled
+    """
+    cdef Py_ssize_t i, j, temp
+
+    for i in range(size-1, 0, -1):
+        j = rand() % (i + 1)
+        temp = arr[i]
+        arr[i] = arr[j]
+        arr[j] = temp
+
+
+@boundscheck(False) # Deactivate bounds checking
+@wraparound(False)  # Deactivate negative indexing
+@cdivision(True) # Ignore modulo/divide by zero warning
+cdef inline void _init_flag_table(UINT8_t* flags, const Py_ssize_t* perm,
+                                  const Py_ssize_t size) nogil:
+    """ Initializes lookup table according to provided permutation
+
+    Args:
+        flags (uint8_t*):
+            pointer to mutable randomized column lookup table memory block
+        perm (Py_ssize_t*):
+            pointer to immutable permutation memory block
+        size (Py_ssize_t):
+            size of the lookup table/permutation
+    """
+    cdef Py_ssize_t idx
+    for idx in range(size):
+        flags[perm[idx]] = 1
+
+
+@boundscheck(False) # Deactivate bounds checking
+@wraparound(False)  # Deactivate negative indexing
+@cdivision(True)    # Ignore modulo/divide by zero warning
+cdef inline void _list_accum(DTYPE_t[:] close_num_rand_view,
+                             const DTYPE_t[:, :] dist_mat_bin, MAXINDEX_t[:] pos_labels,
+                             Py_ssize_t* rand_cols, Py_ssize_t num_choices, int m1n, int m2n,
+                             int bootstrap_num):
+    """ List based accumulation for small secondary marker size
```

Comment:
What is secondary marker size?

Reply:
just the number of cells 'of lineage'/'with marker' `j` (i.e `m2n`).  it's faster than `dict_accum` in the case where `m2n` is smaller than the average `m2n` over all markers/lineages.  I'm open to tweaking the doc string to be more clear.

Reply:
Okay got it. If you're consistent with that I think it's fine to use it